sycl/source/detail/graph_impl.cpp

```diff
@@ -156,6 +146,32 @@ bool isPartitionRoot(std::shared_ptr<node_impl> Node) {
   return true;
 }
 
+/// Takes a vector of weak_ptrs to node_impls and returns a vector of node
+/// objects created from those impls, in the same order.
+std::vector<node> createNodesFromImpls(
+    const std::vector<std::weak_ptr<detail::node_impl>> &Impls) {
+  std::vector<node> Nodes{};
```

Comment:
Best to reserve the number of elements, given we know how many we'll get.
```suggestion
  std::vector<node> Nodes(Impls.size());
```

Reply:
Done! Although with an explicit `reserve()` call since we can't default construct nodes.